[DOCS] Create and Publish an Application Integration Support Policy (#8703)
great-expectations/great_expectations: docs/docusaurus/docs/contributing/application_integration_support.md

| <i class="fas fa-check-circle" style={{color: "#28a745"}}></i> macOS<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Linux    | <i class="fas fa-circle" style={{color: "#dc3545"}}></i> Windows              | 

## Python versions

The following table defines the Python versions supported by GX and the community. GX typically follows the [Python release cycle](https://devguide.python.org/versions/).

| GX supported                             | Community supported                                                |
|------------------------------------------|--------------------------------------------------------------------|
| <i class="fas fa-check-circle" style={{color: "#28a745"}}></i> 3.8 to 3.11    | <i class="fas fa-circle" style={{color: "#dc3545"}}></i> 3.7.X and earlier              | 

## Datasources

The following table defines the Datasources supported by GX and the community.

| GX supported                             | Community supported                                                |
|------------------------------------------|--------------------------------------------------------------------|

> | <i class="fas fa-circle" style={{color: "#dc3545"}}></i> Snowflake<br/><i class="fas fa-circle" style={{color: "#dc3545"}}></i> Databricks<br/><i class="fas fa-circle" style={{color: "#dc3545"}}></i> BigQuery<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> PostgreSQL<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> pandas (maintenance mode)<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Spark (maintenance mode)  | <i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Redshift<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> MSSQL<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> MySQL<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> SQLite<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Trino<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> AWS S3<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Google Cloud Storage<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Azure Blog Storage<br/><i class="fas fa-circle" style={{color: "#ffc107"}}></i> Clickhouse<br/><i class="fas fa-circle" style={{color: "#ffc107"}}></i> Athena<br/><i class="fas fa-circle" style={{color: "#ffc107"}}></i> Dremio<br/><i class="fas fa-circle" style={{color: "#ffc107"}}></i> Teradata<br/><i class="fas fa-circle" style={{color: "#ffc107"}}></i> Vertica<br/><i class="fas fa-circle" style={{color: "#dc3545"}}></i> EMR Spark<br/><i class="fas fa-circle" style={{color: "#dc3545"}}></i> AWS Glue             | 

Reviewer: Should Databricks, Snowflake, and Bigquery be yellow here? Snowflake and BigQuery are certainly under partial test, and we support all three of them.

| GX supported                             | Community supported                                                |
|------------------------------------------|--------------------------------------------------------------------|
| <i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Slack<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Email     | <i class="fas fa-check-circle" style={{color: "#28a745"}}></i> MS Teams<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Pager Duty<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Opsgenie<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Amazon SNS<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> General API<br/><i class="fas fa-circle" style={{color: "#dc3545"}}></i> DataHub        | 

## Credential stores

The following table defines the credential stores supported by GX and the community.

| GX supported                             | Community supported                                                |
|------------------------------------------|--------------------------------------------------------------------|

> | <i class="fas fa-check-circle" style={{color: "#28a745"}}></i> YAML<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> AWS Secrets Store<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Azure Blob Store<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> GCP Secrets Manager<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Azure Key Vault<br/><i class="fas fa-check-circle" style={{color: "#28a745"}}></i> Databricks Secrets Store<br/><i class="fas fa-circle" style={{color: "#dc3545"}}></i> HashiCorp Vault    |               |

Reviewer: I'm thinking HashiCorp shouldn't be under GX supported.

Reviewer: Also uncertain that all of these are under test -- certainly we don't have example scripts showing these under test for Databricks Secrets Store or Azure Blob Store.

---
title: Integration support policy
---

> A defined integration support policy allows Great Expectations (GX) to prioritize its response to application integration, operating system, and programming language issues.

Reviewer 2: ```suggestion
```

> For production environments where uptime is critical, GX recommends using the applications, operating systems, and programming languages defined as GX-supported. Community-supported platforms such as Pandas, Spark, and SQLAlchemy allow you to integrate your existing datasets with GX Cloud and GX OSS.

Reviewer 2: ```suggestion
For production environments, GX recommends using the integrations defined as GX-supported. 

GX uses libraries such as Pandas, Spark, and SQLAlchemy to integrate with a wide range of data sources. This also allows you to deploy GX with Community-supported integrations.
```

| GX Cloud                     | GX OSS                        |
|------------------------------|-------------------------------|
| Mac/Linux ¹                  | Mac/Linux                     | 

¹ Required to run the GX Agent.

### Python versions

The following table defines the Python versions supported by GX Cloud and GX OSS. GX typically follows the [Python release cycle](https://devguide.python.org/versions/).

| GX Cloud                    | GX OSS                          |
|-----------------------------|---------------------------------|
| N/A                         | 3.8 to 3.11                     | 

### GX versions

The following table defines the GX versions supported by GX Cloud and GX OSS.

| GX Cloud                    | GX OSS                          |
|-----------------------------|---------------------------------|
| N/A                         | - 0.17<br/> - 0.18              | 

### Integrations

The following table defines the supported GX Cloud and GX OSS integrations.

| Application Type                         | GX Cloud             | GX OSS          |
|------------------------------------------|----------------------|-----------------|

> | Data Source                              | Snowflake           | - Snowflake<br/> - Generic SQL ¹               |

Reviewer 2: Support for Bigquery coming soon.

### GX components

The following table defines the GX components supported by GX Cloud and GX OSS.

| Component                                | GX Cloud             | GX OSS          |
|------------------------------------------|----------------------|-----------------|
| Expectations                             | See [Available Expectations](/docs/cloud/expectations/manage_expectations#available-expectations). | See [Create Expectations](/docs/guides/expectations/expectations_lp).                |
| GX Agent                                 | All versions               | N/A        |

## Community supported

The following integrated applications, operating systems, and programming languages are supported by the community.

### Operating systems

The following table defines the operating systems supported by the community.

| GX Cloud                       | GX OSS                        |
|--------------------------------|-------------------------------|
| Windows ¹                      | Windows ¹                     | 

¹ Untested and unsupported by GX.

### Integrations

The following table defines the GX Cloud and GX OSS integrations supported by the community.

> | Application Type                         | GX Cloud             | GX OSS          |

Reviewer 2: All GX Cloud is N/A.

Explain why N/A. All Community-supported integrations are applicable only to GX OSS.

|------------------------------------------|----------------------|-----------------|
| Data Source                              | Snowflake          | - Redshift<br/>- MSSQL<br/>- MySQL<br/>- SQLite<br/>- Trino<br/>- AWS S3<br/>- Google Cloud Storage<br/>- Azure Blog Storage<br/>- Clickhouse ¹<br/>- Athena ¹<br/>- Dremio ¹<br/>- Teradata ¹<br/>- Vertica ¹<br/>- EMR Spark ²<br/>- AWS Glue ²               |
| Configuration Stores                     | N/A                | Azure Blob Store                |
| Communication (notifications)            | N/A                | - Slack (local) ³<br/> - email (local)<br/> - Microsoft Teams (local)<br/> - PagerDuty (local)<br/>- Opsgenie<br/>- Amazon SNS<br/>- General API<br/>- DataHub ⁴                   |
| Orchestrator                              | Airflow ⁵           | - Airflow<br/>- Prefect<br/>- Dagster ⁶<br/>- Flyte ⁶<br/>- mage.ai ⁶<br/>- Github Action ⁶        |

> ¹ Partially tested and supported by GX.<br/>

Reviewer 2: ```suggestion
```

> ² Untested and unsupported by GX.<br/>

Reviewer 2: delete

---

> title: Application support policy

Reviewer 2: ```suggestion
title: Support policy
```

Author: @joshzzheng - support policy for what? You need to specify the type of policy. This is a policy for application integrations, operating systems, and programming languages.

---

> A defined application support policy allows Great Expectations (GX) to better prioritize its response to application integration, operating system, and programming language issues. Formally defining the support policy also provides contributors with a better understanding of who is responsible for resolving application integration, operating system, and programming language issues.

Reviewer 2: ```suggestion
A defined support policy allows Great Expectations (GX) to prioritize its response to application integration, operating system, and programming language issues.

For production environments where uptime is critical, GX recommends using the applications, operating systems, and programming languages defined as GX-supported. Community-supported platforms such as Pandas, Spark, and SQLAlchemy allow you to integrate your existing datasets with GX Cloud and GX OSS.
```

## Support categories

> The following are the types of support available for application integration, operating system, and programming language issues:

Reviewer 2: ```suggestion
The following are the levels of support provided by GX:

- GX-supported - integrations are tested throughout the development lifecycle, and actively maintained and updated when new versions of GX Cloud or GX OSS are released.

- Community-supported - integrations are successfully implemented by community members. They are not tested or updated by GX, and GX is not responsible for ensuring reliability or compatibility.
```

> - **GX supported** - The GX Developer Relations (DevRel) team triages, reproduces, and then assigns issues to the appropriate GX Engineering team. Issues with application integration, operating system, and programming languages in maintenance mode are reviewed, but new features or functionality are not implemented as part of the resolution.

Reviewer 2: ```suggestion
```

> - **Community supported** - The GX DevRel team triages and reproduces issues, and then determines if GX should implement new features or functionality to resolve the issue. If the DevRel team determines that new features or functionality are not warranted, the GX community is responsible for issue resolution. If new features or functionality are required, DevRel assigns the issue to the appropriate GX Engineering team.

Reviewer 2: ```suggestion
```

## GX supported

> The following are the levels of support offered by GX for integrated applications, operating systems, and programming languages.

Reviewer 2: ```suggestion
```

Author: @joshzzheng, we can't use empty headings. See heading and title format in the Google developer documentation style guide (https://developers.google.com/style/headings#heading-and-title-format). We need an introductory statement after the heading.

### Operating systems

The following table defines the operating systems supported by GX Cloud and GX OSS.

| GX Cloud                     | GX OSS                        |
|------------------------------|-------------------------------|

> | - Mac/Linux ¹                | - Mac/Linux                   | 

Reviewer 2: ```suggestion
 Mac/Linux ¹                | Mac/Linux                   | 
```

> | - Mac/Linux ¹                | - Mac/Linux                   | 

Reviewer 2: I'd like to remove these bullets for single items, I think they make things harder to read,

|-----------------------------|---------------------------------|
| - N/A                       | - 0.17<br/> - 0.18              | 

### Integrations

The following table defines the supported GX Cloud and GX OSS integrations.

| Application Type                         | GX Cloud             | GX OSS          |
|------------------------------------------|----------------------|-----------------|
| Data Source                              | - Snowflake          | - Snowflake<br/> - Generic SQL ¹               |
| Configuration Stores                     | - N/A                | - Filesystem<br/> - AWS S3GCS Buckets<br/> - Azure Blob Store                |
| Communication (notifications)            | - N/A ²               | - Slack (local) ³<br/> - email (local)<br/> - Microsoft Teams (local)<br/> - PagerDuty (local)                 |
| Orchestrator                              | - Airflow ⁴           | - Airflow ⁴       |

¹ Connection strings for various SQL Data Sources are supported, but dialect-specific SQL commands are not.<br/>
² Support for Slack, email, and Zapier are planned for future releases.<br/>
³ When used with the GX OSS library.<br/>

> ⁴ Other Orchestrator applications that execute Python code are not supported.

Reviewer 2: ```suggestion
⁴ GX officially only supports Airflow, but it should work with any orchestrator that executes Python code.
```

### GX components

The following table defines the GX components supported by GX Cloud and GX OSS.

| Component                                | GX Cloud             | GX OSS          |
|------------------------------------------|----------------------|-----------------|
| Expectations                             | - See [Available Expectations](/docs/cloud/expectations/manage_expectations#available-expectations) ¹. | - See [Create Expectations](/docs/guides/expectations/expectations_lp) ¹.                |
| GX Agent                                 | - All versions               | - N/A        |

> ¹ The creation and execution of custom Expectations are supported, but inner queries are not.

Reviewer 2: ```suggestion
```

Author: @joshzzheng, we can't use empty headings. See heading and title format in the Google developer documentation style guide (https://developers.google.com/style/headings#heading-and-title-format). We need an introductory statement after the heading.

## Community supported

> The following integrated applications, operating systems, and programming languages are supported by the community.

Reviewer 2: ```suggestion
```

I dont think we need this line. We already explained the two tiers above.

Author: @joshzzheng, we can't use empty headings. See heading and title format in the Google developer documentation style guide (https://developers.google.com/style/headings#heading-and-title-format). We need an introductory statement after the heading.

### Operating systems

The following table defines the operating systems supported by the community.

| GX Cloud                       | GX OSS                        |
|--------------------------------|-------------------------------|

> | - Mac/Linux ¹<br/>- Windows ²  | - Mac/Linux<br/>- Windows ²   | 

Reviewer 2: ```suggestion
| - Windows ²  | Windows ²   | 
```

Mac/Linux is GX-supported not community supported

¹ Required to run the GX Agent.<br/>
² Untested and unsupported by GX

> ### Python versions

Reviewer 2: We're not support previous versions at all. We should remove both Python versions and GX versions section.

---
title: Integration support policy
---

For production environments, GX recommends using the integrations defined as GX-supported. 

GX uses libraries such as Pandas, Spark, and SQLAlchemy to integrate with a wide range of data sources. This also allows you to deploy GX with Community-supported integrations.

## Support categories

The following are the levels of support provided by GX:

> - GX-supported - integrations are tested throughout the development lifecycle, and actively maintained and updated when new versions of GX Cloud or GX OSS are released.

Reviewer 2: ```suggestion
- GX-supported - integrations are tested throughout the development lifecycle, and are actively maintained and updated when new versions of GX Cloud or GX OSS are released.
```

> - Community-supported - integrations are implemented by community members. They are not tested or updated by GX, and GX is not responsible for ensuring reliability or compatibility.

Reviewer 2: These integrations are are not just implemented by community members. Some of these were created by GX. We also test and update some of them. We've also seen successful implementations of some of these. I'm ok with the last piece around responsible for reliability or compatibility.

|------------------------------------------|----------------------|-----------------|
| Expectations                             | See [Available Expectations](/docs/cloud/expectations/manage_expectations#available-expectations). | See [Create Expectations](/docs/guides/expectations/expectations_lp).                |
| GX Agent                                 | All versions               | N/A        |

## Community supported

The following integrated applications, operating systems, and programming languages are supported by the community.

### Operating systems

The following table lists the operating systems supported by the community.

| GX Cloud                       | GX OSS                        |
|--------------------------------|-------------------------------|
| N/A                            | Windows ¹                     | 

¹ Untested and unsupported by GX.

### Integrations

The following table lists the GX Cloud and GX OSS integrations supported by the community.

| Integration Type                         | GX Cloud             | GX OSS          |
|------------------------------------------|----------------------|-----------------|

> | Data Source                              | N/A          | Pandas<br/>Spark<br/>Databricks (Spark)<br/>Databricks (SQL)<br/>Trino<br/>Clickhouse<br/>Dremio<br/> Teradata<br/>Vertica<br/>EMR Spark<br/>AWS Glue<br/>Google Cloud Storage<br/>Azure Blog Storage<br/>AWS S3|

Reviewer 2: ```suggestion
| Data Sources                              | N/A          | Pandas<br/>Spark<br/>Databricks (Spark)<br/>Databricks (SQL)<br/>Trino<br/>Clickhouse<br/>Dremio<br/> Teradata<br/>Vertica<br/>EMR Spark<br/>AWS Glue<br/>Google Cloud Storage<br/>Azure Blog Storage<br/>AWS S3|
```

| Notifications                             | N/A            | Opsgenie<br/>Amazon SNS<br/>DataHub |

> | Orchestrator                              | N/A            | Prefect<br/>Dagster <br/>Flyte <br/>mage.ai  |

Reviewer 2: ```suggestion
| Orchestrators                             | N/A            | Prefect<br/>Dagster <br/>Flyte <br/>mage.ai  |
```
